nrf52-hal-common/src/uarte.rs

```diff
@@ -359,9 +340,530 @@ impl<T> Uarte<T> where T: Instance {
+        timer: Timer<I>,
+        dma_pool_memory: &'static mut [u8],
+    ) -> (UarteRX<T, I>, UarteTX<T, S>)
+    where
+        S: ArrayLength<heapless::pool::singleton::Box<UarteDMAPool>>,
+        I: timer::Instance,
+    {
+        debug_assert!(
+            dma_pool_memory.len() >= core::mem::size_of::<UarteDMAPoolNode>() * 2,
+            "The memory pool needs at least space for 2 `UarteDMAPoolNode`s"
+        );
+        UarteDMAPool::grow(dma_pool_memory);
+
+        let rx = UarteRX::<T, I>::new(timer);
+        let tx = UarteTX::<T, S>::new(txc);
+        (rx, tx)
+    }
 }
 
-impl<T> fmt::Write for Uarte<T> where T: Instance {
+/// An interrupt driven implementation of the UARTE peripheral.
+///
+/// ## Example
+///
+/// ```
+///
+/// ```
+pub mod interrupt_driven {
+    use core::sync::atomic::{compiler_fence, Ordering::SeqCst};
+    use core::{fmt, mem, ptr, slice};
+
+    use crate::timer::{self, Timer};
+    use crate::uarte;
+    use embedded_hal::timer::{Cancel, CountDown};
+
+    use heapless::{
+        consts, pool,
+        pool::singleton::{Box, Pool},
+        spsc::{Consumer, Queue},
+        ArrayLength,
+    };
+
+    /// DMA block size
+    /// Defaults to UARTE_DMA_SIZE = 16 if not explicitly set
+    #[cfg(not(any(
+        feature = "UARTE_DMA_SIZE_4",
+        feature = "UARTE_DMA_SIZE_8",
+        feature = "UARTE_DMA_SIZE_16",
+        feature = "UARTE_DMA_SIZE_32",
+        feature = "UARTE_DMA_SIZE_64",
+        feature = "UARTE_DMA_SIZE_128",
+        feature = "UARTE_DMA_SIZE_255"
+    )))]
+    pub const UARTE_DMA_SIZE: usize = 16;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_4")]
+    pub const UARTE_DMA_SIZE: usize = 4;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_8")]
+    pub const UARTE_DMA_SIZE: usize = 8;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_16")]
+    pub const UARTE_DMA_SIZE: usize = 16;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_32")]
+    pub const UARTE_DMA_SIZE: usize = 32;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_64")]
+    pub const UARTE_DMA_SIZE: usize = 64;
+
+    #[cfg(feature = "UARTE_DMA_SIZE_128")]
+    pub const UARTE_DMA_SIZE: usize = 128;
+
+    // Currently causes internal OOM, needs fixing
+    // Maximum DMA size is 255 of the UARTE peripheral, see `MAXCNT` for details
+    #[cfg(feature = "UARTE_DMA_SIZE_255")]
+    pub const UARTE_DMA_SIZE: usize = 255;
+
+    // An alternative solution to the above is to define the UARTE_DMA_SIZE
+    // in a separate (default) crate, which can be overridden
+    // by a patch in the user Cargo.toml, pointing to
+    // a local crate with the user defined UARTE_DMA_SIZE constant.
+    // What would you prefer?
+
+    pool!(
+        /// Pool allocator for the DMA engine running the transfers, where each node in the
+        /// allocator is an `UarteDMAPoolNode`
+        #[allow(non_upper_case_globals)]
+        UarteDMAPool: UarteDMAPoolNode
+    );
+
+    /// Each node in the `UarteDMAPool` consists of this struct.
+    pub struct UarteDMAPoolNode {
```

Comment:
Should be `UartDmaPoolNode` as per established convention. I'm all for ignoring conventions when there's a good reason, but in this case, I think the conventional name is easier to parse.